src/pages/[lang]/404.astro

```diff
@@ -1,11 +1,11 @@
 ---
-import languages from '../../i18n/languages';
+import languages from '~/i18n/languages';
+import { getLanguageFromURL } from '~/util';
 import { useTranslations } from '../../i18n/util';
 import SplashLayout from '../../layouts/SplashLayout.astro';
 
 export const getStaticPaths = () => Object.keys(languages).map(lang => ({ params: { lang } }));
-
-const { lang } = Astro.params;
+const lang = getLanguageFromURL(Astro.canonicalURL.pathname) || 'en';
```

Comment:
The `lang` is available in params everywhere except on the root 404, so I think we can just add a fallback value here and avoid importing `getLanguageFromURL`.

```suggestion
const { lang = 'en' } = Astro.params;
```